doc/architecture/decisions/0003-use-open-api-generator-to-scaffold-controllers-from-open-api-file.md

```diff
@@ -0,0 +1,52 @@
+# 3. Use Open API Generator to scaffold Controllers from Open API file
+
+Date: 2022-08-03
+
+## Status
+
+Proposed
+
+## Context
+
+To provide API documentation in an industry standard way we have already decided to use an OpenAPI file.  Doing so 
+adds the burden of ensuring that the actual API implementation aligns with what is documented in the file.
+
+## Decision
+
+We will use [OpenAPI Generator](https://github.com/OpenAPITools/openapi-generator) to automate the process of 
+creating Spring Controllers.  By making the `compileKotlin` gradle Task depend on the `openApiGenerate` task this 
+will happen automatically each time the project is built.
+
+## Consequences
+
+This reduces repetitive work and ensures that the OpenAPI file and the actual endpoints always align.  It also opens 
+the possibility of generating other types of code in the future to reduce repetitive work for clients of the API:
+ - [TypeScript Client](https://openapi-generator.tech/docs/generators/typescript-fetch)
+ - [Kotlin Client](https://openapi-generator.tech/docs/generators/kotlin)
+
+
+Unfortunately, OpenAPI Generator does not support the OpenAPI 3 `anyOf`, `oneOf` options on request/response bodies 
+which are useful if you want to return completely different responses from an endpoint.  Whilst this is on their 
+[roadmap](https://openapi-generator.tech/docs/roadmap/#short-term) it has been there for some time.  The implication of 
+this is that if you want to return either a success or error response from the same endpoint then you need to do so in 
+an enveloped response, e.g.:
```

Comment:
Does this mean we need to make changes to the OpenAPI schema? Could we do this in this PR, so we don't forget?

Reply:
Quite possibly, do we have an idea of the what the shape of the error responses are going to be or is this TBD?

Reply:
I've just spotted the ADR for problem responses, so all good - will have a read.

Reply:
There's also an example here https://github.com/ministryofjustice/approved-premises-api/blob/main/src/main/resources/static/mini-manage-api-stubs.yml#L156

Reply:
So this works out quite actually and seems we won't need to do the envelope response after all.

Whilst Open API Generate doesn't support returning different models from the delegate methods (it just picks the first one from the possible responses, which should always be the 2xx response) this won't be a problem since the https://github.com/zalando/problem-spring-web library uses exceptions to convey a failure - the exception handlers aren't bound by the same restrictions in what can be returned.

I have added the problem responses to the other 4/5xx responses in the spec too.

Reply:
Have also updated the ADR to reflect this finding